File: 043/cpp/main.cpp
class Solution {
public:
    string multiply1(string num1, char p) {
        int q = 0;
        string s;
        for (int i = num1.size() - 1; i > -1; i--) {
            q += (num1[i] - '0') * (p - '0');
            s = string(1, q % 10 + '0') + s;
            q /= 10;
        }
        if (q > 0) {
            s = string(1, q + '0') + s;
        }
        return s;
    }

    string add(string num1, string num2) {
        int q = 0;
        int n1 = num1.size();
        int n2 = num2.size();
        int n = max(n1, n2);
        string s;
        for (int i = 0; i < n; i++) {
            char a = n1 >= i + 1 ? num1[n1 - 1 - i] : '0';
            char b = n2 >= i + 1 ? num2[n2 - 1 - i] : '0';
            q += ((a - '0') + (b - '0'));
            s = string(1, q % 10 + '0') + s;
            q /= 10;
        }
        if (q > 0) {
            s = string(1, q + '0') + s;
        }
        return s;
    }

    string multiply(string num1, string num2) {
        if (num1 == "0" || num2 == "0") {
            return "0";
        }
        string s2 = multiply1(num1, num2[num2.size() - 1]);
        if (num2.size() == 1) {
            return s2;
        }
        string s1 = multiply(num1, num2.substr(0, num2.size() - 1));
        return add(s1 + "0", s2);
    }
};

File: 043/rust/main.rs
use std::cmp;

impl Solution {
    fn multiply1(num1: &[u8], p: u8) -> String {
        let p = (p as char).to_digit(10).unwrap();
        let mut q = 0;
        let mut s = String::new();
        for b in num1.iter().rev() {
            let c = (*b as char).to_digit(10).unwrap();
            q += c * p; 
            s = format!("{}{}", q % 10, s);
            q /= 10; 
        }
        if q > 0 {
            s = format!("{}{}", q, s);
        }
        return s;
    }

    fn multiply2(num1: &[u8], num2: &[u8]) -> String {
        let p = num2[num2.len() - 1];
        let s2 = Solution::multiply1(num1, p);
        if num2.len() == 1 {
            return s2;
        }
        let s1 = Solution::multiply2(num1, &num2[0..num2.len() - 1]);
        let s1 = format!("{}0", s1);
        return Solution::add(s1.as_bytes(), s2.as_bytes());
    }

    fn add(num1: &[u8], num2: &[u8]) -> String {
        let n1 = num1.len();
        let n2 = num2.len();
        let n = cmp::max(n1, n2);
        let mut q = 0;
        let mut s = String::new();
        for i in 0..n {
            let j1 = num1.len() - 1 - i;
            let j2 = num2.len() - 1 - i;
            let a = if n1 >= i + 1 { (num1[n1 - 1 - i] as char).to_digit(10).unwrap() } else { 0 };
            let b = if n2 >= i + 1 { (num2[n2 - 1 - i] as char).to_digit(10).unwrap() } else { 0 };
            q += a + b;
            s = format!("{}{}", q % 10, s);
            q /= 10;
        }
        if q > 0 {
            s = format!("{}{}", q, s);
        }
        return s;
    }

    pub fn multiply(num1: String, num2: String) -> String {
        let zero = "0".to_string();
        if num1 == zero || num2 == zero {
            return zero;
        }
        let num1 = num1.as_bytes();
        let num2 = num2.as_bytes();
        return Solution::multiply2(num1, num2);
    }
}
